File: lib/input-device.mjs
const EV_KEY = 0x01;
const EV_REL = 0x02;
const REL_X = 0x00;
const REL_Y = 0x01;
const REL_WHEEL = 0x08;
const BTN_LEFT = 0x110;
const BTN_RIGHT = 0x111;
const BTN_MIDDLE = 0x112;

const KEY_CODES = new Map(Object.entries({
  Escape: 1,
  Digit1: 2,
  Digit2: 3,
  Digit3: 4,
  Digit4: 5,
  Digit5: 6,
  Digit6: 7,
  Digit7: 8,
  Digit8: 9,
  Digit9: 10,
  Digit0: 11,
  Minus: 12,
  Equal: 13,
  Backspace: 14,
  Tab: 15,
  KeyQ: 16,
  KeyW: 17,
  KeyE: 18,
  KeyR: 19,
  KeyT: 20,
  KeyY: 21,
  KeyU: 22,
  KeyI: 23,
  KeyO: 24,
  KeyP: 25,
  BracketLeft: 26,
  BracketRight: 27,
  Enter: 28,
  ControlLeft: 29,
  ControlRight: 97,
  KeyA: 30,
  KeyS: 31,
  KeyD: 32,
  KeyF: 33,
  KeyG: 34,
  KeyH: 35,
  KeyJ: 36,
  KeyK: 37,
  KeyL: 38,
  Semicolon: 39,
  Quote: 40,
  Backquote: 41,
  ShiftLeft: 42,
  Backslash: 43,
  KeyZ: 44,
  KeyX: 45,
  KeyC: 46,
  KeyV: 47,
  KeyB: 48,
  KeyN: 49,
  KeyM: 50,
  Comma: 51,
  Period: 52,
  Slash: 53,
  ShiftRight: 54,
  AltLeft: 56,
  AltRight: 100,
  Space: 57,
  CapsLock: 58,
  F1: 59,
  F2: 60,
  F3: 61,
  F4: 62,
  F5: 63,
  F6: 64,
  F7: 65,
  F8: 66,
  F9: 67,
  F10: 68,
  F11: 87,
  F12: 88,
  Home: 102,
  ArrowUp: 103,
  PageUp: 104,
  ArrowLeft: 105,
  ArrowRight: 106,
  End: 107,
  ArrowDown: 108,
  PageDown: 109,
  Insert: 110,
  Delete: 111,
  MetaLeft: 125,
  MetaRight: 126
}));

const ASCII_KEY_CODES = new Map([
  ["\n", 28], ["\t", 15], [" ", 57],
  ["0", 11], ["1", 2], ["2", 3], ["3", 4], ["4", 5],
  ["5", 6], ["6", 7], ["7", 8], ["8", 9], ["9", 10],
  ["a", 30], ["b", 48], ["c", 46], ["d", 32], ["e", 18], ["f", 33],
  ["g", 34], ["h", 35], ["i", 23], ["j", 36], ["k", 37], ["l", 38],
  ["m", 50], ["n", 49], ["o", 24], ["p", 25], ["q", 16], ["r", 19],
  ["s", 31], ["t", 20], ["u", 22], ["v", 47], ["w", 17], ["x", 45],
  ["y", 21], ["z", 44],
  ["-", 12], ["=", 13], ["[", 26], ["]", 27], ["\\", 43],
  [";", 39], ["'", 40], ["`", 41], [",", 51], [".", 52], ["/", 53],
  ["!", 2], ["@", 3], ["#", 4], ["$", 5], ["%", 6],
  ["^", 7], ["&", 8], ["*", 9], ["(", 10], [")", 11],
  ["_", 12], ["+", 13], ["{", 26], ["}", 27], ["|", 43],
  [":", 39], ['"', 40], ["~", 41], ["<", 51], [">", 52], ["?", 53],
]);

const SHIFT_CHARS = new Set('!@#$%^&*()_+{}|:"<>?ABCDEFGHIJKLMNOPQRSTUVWXYZ');

export class InputDevice {
  constructor(adapter) {
    this.adapter = adapter;
  }

  move(dx, dy) {
    this.adapter.move(dx, dy);
  }

  click(button) {
    this.adapter.click(button);
  }

  key(code, down) {
    this.adapter.key(code, down);
  }

  type(text) {
    this.adapter.type(text);
  }

  scroll(dy) {
    this.adapter.scroll(dy);
  }
}

export class UinputAdapter {
  constructor(stream) {
    this.stream = stream;
  }

  move(dx, dy) {
    this._emit(EV_REL, REL_X, dx, false);
    this._emit(EV_REL, REL_Y, dy, true);
  }

  click(button) {
    const btnMap = { left: BTN_LEFT, right: BTN_RIGHT, middle: BTN_MIDDLE };
    const code = btnMap[button] ?? BTN_LEFT;
    this._emit(EV_KEY, code, 1, true);
    this._emit(EV_KEY, code, 0, true);
  }

  key(code, down) {
    const keyCode = KEY_CODES.get(code);
    if (keyCode !== undefined) {
      this._emit(EV_KEY, keyCode, down ? 1 : 0, true);
    }
  }

  type(text) {
    for (const ch of text) {
      const code = ASCII_KEY_CODES.get(ch);
      if (code === undefined) continue;
      const needShift = SHIFT_CHARS.has(ch);
      if (needShift) {
        this._emit(EV_KEY, 42, 1, true);
      }
      this._emit(EV_KEY, code, 1, true);
      this._emit(EV_KEY, code, 0, true);
      if (needShift) {
        this._emit(EV_KEY, 42, 0, true);
      }
    }
  }

  scroll(dy) {
    this._emit(EV_REL, REL_WHEEL, dy, true);
  }

  _emit(type, code, value, synReport = false) {
    const buf = Buffer.alloc(24);
    buf.writeUInt16LE(type, 16);
    buf.writeUInt16LE(code, 18);
    buf.writeInt32LE(value, 20);
    this.stream.write(buf);
    if (synReport) {
      buf.writeUInt16LE(0x00, 16);
      buf.writeUInt16LE(0x00, 18);
      buf.writeInt32LE(0, 20);
      this.stream.write(buf);
    }
  }
}

File: macos/TapPad/Sources/TapPad/Input/InputDevice.swift
import AppKit
import CoreGraphics
import Foundation

protocol InputDevice {
    func move(dx: Double, dy: Double)
    func click(button: String, clickCount: Int)
    func scroll(dy: Double)
    func key(code: String, down: Bool)
    func typeText(_ text: String)
}

final class MacInputDevice: InputDevice, @unchecked Sendable {
    private let lock = NSLock()
    private let scrollPoster = ScrollPoster()
    private var cursor: CGPoint
    private let desktopBounds: CGRect

    init() {
        desktopBounds = Self.activeDesktopBounds()
        cursor = CGEvent(source: nil)?.location
            ?? CGPoint(x: desktopBounds.midX, y: desktopBounds.midY)
    }

    func move(dx: Double, dy: Double) {
        lock.withLock {
            cursor.x = min(max(cursor.x + dx, desktopBounds.minX), desktopBounds.maxX - 1)
            cursor.y = min(max(cursor.y + dy, desktopBounds.minY), desktopBounds.maxY - 1)
            postMouse(.mouseMoved, at: cursor, button: .left)
        }
    }

    func click(button: String, clickCount: Int) {
        lock.withLock {
            let kind = mouseKind(button)
            postMouse(kind.down, at: cursor, button: kind.button, clickCount: clickCount)
            postMouse(kind.up, at: cursor, button: kind.button, clickCount: clickCount)
        }
    }

    func scroll(dy: Double) {
        scrollPoster.push(deltaY: dy)
    }

    func key(code: String, down: Bool) {
        guard let keyCode = keyCode(for: code),
              let event = CGEvent(keyboardEventSource: nil, virtualKey: keyCode, keyDown: down)
        else {
            return
        }
        event.post(tap: .cghidEventTap)
    }

    func typeText(_ text: String) {
        for scalar in text.unicodeScalars {
            var utf16 = Array(String(scalar).utf16)
            postUnicodeKey(&utf16)
        }
    }

    private func postMouse(
        _ type: CGEventType,
        at point: CGPoint,
        button: CGMouseButton,
        clickCount: Int = 1
    ) {
        guard let event = CGEvent(
            mouseEventSource: nil,
            mouseType: type,
            mouseCursorPosition: point,
            mouseButton: button
        ) else {
            return
        }
        event.setIntegerValueField(.mouseEventClickState, value: Int64(clickCount))
        event.post(tap: .cghidEventTap)
    }

    private static func activeDesktopBounds() -> CGRect {
        var displayCount: UInt32 = 0
        guard CGGetActiveDisplayList(0, nil, &displayCount) == .success, displayCount > 0 else {
            return CGDisplayBounds(CGMainDisplayID())
        }

        var displays = [CGDirectDisplayID](repeating: 0, count: Int(displayCount))
        guard CGGetActiveDisplayList(displayCount, &displays, &displayCount) == .success else {
            return CGDisplayBounds(CGMainDisplayID())
        }

        return displays
            .prefix(Int(displayCount))
            .map(CGDisplayBounds)
            .reduce(CGRect.null) { $0.union($1) }
    }

    private func postUnicodeKey(_ utf16: inout [UniChar]) {
        guard let down = CGEvent(keyboardEventSource: nil, virtualKey: 0, keyDown: true) else {
            return
        }
        down.keyboardSetUnicodeString(stringLength: utf16.count, unicodeString: &utf16)
        down.post(tap: .cghidEventTap)

        guard let up = CGEvent(keyboardEventSource: nil, virtualKey: 0, keyDown: false) else {
            return
        }
        up.keyboardSetUnicodeString(stringLength: utf16.count, unicodeString: &utf16)
        up.post(tap: .cghidEventTap)
    }

    private func mouseKind(_ name: String) -> (
        down: CGEventType,
        up: CGEventType,
        button: CGMouseButton
    ) {
        switch name {
        case "right":
            return (.rightMouseDown, .rightMouseUp, .right)
        case "middle":
            return (.otherMouseDown, .otherMouseUp, .center)
        default:
            return (.leftMouseDown, .leftMouseUp, .left)
        }
    }

    private func keyCode(for name: String) -> CGKeyCode? {
        switch name {
        case "KeyA": 0
        case "KeyS": 1
        case "KeyD": 2
        case "KeyF": 3
        case "KeyH": 4
        case "KeyG": 5
        case "KeyZ": 6
        case "KeyX": 7
        case "KeyC": 8
        case "KeyV": 9
        case "KeyB": 11
        case "KeyQ": 12
        case "KeyW": 13
        case "KeyE": 14
        case "KeyR": 15
        case "KeyY": 16
        case "KeyT": 17
        case "Digit1": 18
        case "Digit2": 19
        case "Digit3": 20
        case "Digit4": 21
        case "Digit6": 22
        case "Digit5": 23
        case "Equal": 24
        case "Digit9": 25
        case "Digit7": 26
        case "Minus": 27
        case "Digit8": 28
        case "Digit0": 29
        case "BracketRight": 30
        case "KeyO": 31
        case "KeyU": 32
        case "BracketLeft": 33
        case "KeyI": 34
        case "KeyP": 35
        case "Enter": 36
        case "KeyL": 37
        case "KeyJ": 38
        case "Quote": 39
        case "KeyK": 40
        case "Semicolon": 41
        case "Backslash": 42
        case "Comma": 43
        case "Slash": 44
        case "KeyN": 45
        case "KeyM": 46
        case "Period": 47
        case "Tab": 48
        case "Space": 49
        case "Backquote": 50
        case "Backspace": 51
        case "Escape": 53
        case "MetaLeft", "MetaRight": 55
        case "ShiftLeft": 56
        case "CapsLock": 57
        case "AltLeft", "AltRight": 58
        case "ControlLeft", "ControlRight": 59
        case "ShiftRight": 60
        case "F1": 122
        case "F2": 120
        case "F3": 99
        case "F4": 118
        case "F5": 96
        case "F6": 97
        case "F7": 98
        case "F8": 100
        case "F9": 101
        case "F10": 109
        case "F11": 103
        case "F12": 111
        case "Home": 115
        case "PageUp": 116
        case "Delete": 117
        case "End": 119
        case "PageDown": 121
        case "ArrowLeft": 123
        case "ArrowRight": 124
        case "ArrowDown": 125
        case "ArrowUp": 126
        default: nil
        }
    }
}
